reducers/todos.js

```diff
@@ -21,66 +22,61 @@ function todo(state, action) {
       return {
         text: action.text,
         id: action.id,
-        relatedId: null
+        relatedId: null,
+        isCompleted: false
       }
     case EDIT_TODO:
       return {
         ...state,
         text: action.text
       }
+    case COMPLETE_TODO:
+      return {
+        ...state,
+        isCompleted: !state.isCompleted
+      }
   }
 }
 
-function byId(state = initialById, action) {
+function byId(state = initialById, action, {listedIds}) {
   switch (action.type) {
     case ADD_TODO:
     case EDIT_TODO:
+    case COMPLETE_TODO:
       return {
         ...state,
         [action.id]: todo(state[action.id], action)
       }
+    case COMPLETE_ALL:
+      const areSomeActive = listedIds.some(id => !state[id].isCompleted)
+      return Object.keys(state).reduce((nextState, id) => {
+        nextState[id] = {
+          ...state[id],
+          isCompleted: areSomeActive
+        }
+        return nextState
+      }, {})
     default:
       return state
   }
 }
 
-function listedIds(state = initialListedIds, action, {isCompletedById}) {
+function listedIds(state = initialListedIds, action, {byId}) {
   switch (action.type) {
     case ADD_TODO:
       return [action.id, ...state]
     case DELETE_TODO:
-      return state.filter(todoId => todoId !== action.id)
+      return state.filter(id => id !== action.id)
     case CLEAR_COMPLETED:
-      return state.filter(id => !isCompletedById[id])
+      return state.filter(id => !byId[id].isCompleted)
     default:
       return state
   }
 }
 
-function isCompletedById(state = {}, action, {listedIds}) {
-  switch (action.type) {
-    case COMPLETE_TODO:
-      return {
-        ...state,
-        [action.id]: !state[action.id]
-      }
-    case COMPLETE_ALL:
-      const areAllCompleted = listedIds.every(id => state[id])
-      if (areAllCompleted) {
-        return {}
-      }
-      let nextState = {}
-      listedIds.forEach(id => nextState[id] = true)
-      return nextState
-    default:
-       return state
-  }
-}
-
 export default function todos(state = {}, action) {
   return {
-    byId: byId(state.byId, action),
-    listedIds: listedIds(state.listedIds, action, state),
-    isCompletedById: isCompletedById(state.isCompletedById, action, state)
+    byId: byId(state.byId, action, state),
```

Comment:
Yeah, it’s more of a traditional “computer science” approach (lol I never learned computer science though). Meaning: think about who reads this data and how often. Think about who changes this data and how often. Think whether you want to make the reads fast or the writes fast. Think what can be memoized. That’s it.